hibernate-core/src/main/java/org/hibernate/dialect/Dialect.java

```diff
@@ -556,6 +555,64 @@ public String extractPattern(TemporalUnit unit) {
 	 *           type the value argument is cast to
 	 */
 	public String castPattern(CastType from, CastType to) {
+		switch ( to ) {
+			case STRING:
+				switch ( from ) {
+					case INTEGER_BOOLEAN:
+						return "case ?1 when 1 then 'true' when 0 then 'false' else null end";
+					case YN_BOOLEAN:
+						return "case ?1 when 'Y' then 'true' when 'N' then 'false' else null end";
+					case TF_BOOLEAN:
+						return "case ?1 when 'T' then 'true' when 'F' then 'false' else null end";
+				}
+				break;
+			case INTEGER_BOOLEAN:
+				switch ( from ) {
+					case STRING:
+						return "case ?1 when 'T' then 1 when 'Y' then 1 when 'F' then 0 when 'N' then 0 else null end";
+					case YN_BOOLEAN:
+						return "case ?1 when 'Y' then 1 when 'N' then 0 else null end";
+					case TF_BOOLEAN:
+						return "case ?1 when 'T' then 1 when 'F' then 0 else null end";
+					case BOOLEAN:
+						return "case ?1 when true then 1 when false then 0 else null end";
+				}
+			case YN_BOOLEAN:
+				switch ( from ) {
+					case STRING:
+						return "case ?1 when 'T' then 'Y' when 'Y' then 'Y' when 'F' then 'N' when 'N' then 'N' else null end";
+					case INTEGER_BOOLEAN:
+						return "case ?1 when 1 then 'Y' when 0 then 'N' else null end";
+					case TF_BOOLEAN:
+						return "case ?1 when 'T' then 'Y' when 'F' then 'N' else null end";
+					case BOOLEAN:
+						return "case ?1 when true then 'Y' when false then 'N' else null end";
+				}
+			case TF_BOOLEAN:
+				switch ( from ) {
+					case STRING:
+						return "case ?1 when 'T' then 'T' when 'Y' then 'T' when 'F' then 'F' when 'N' then 'F' else null end";
+					case INTEGER_BOOLEAN:
+						return "case ?1 when 1 then 'T' when 0 then 'F' else null end";
+					case YN_BOOLEAN:
+						return "case ?1 when 'Y' then 'T' when 'N' then 'F' else null end";
+					case BOOLEAN:
+						return "case ?1 when true then 'T' when false then 'F' else null end";
+				}
+			case BOOLEAN:
+				switch ( from ) {
+					case STRING:
+						return "case ?1 when 'T' then true when 'Y' then true when 'F' then false when 'N' then false else null end";
+					case INTEGER:
+					case INTEGER_BOOLEAN:
+					case LONG:
+						return "(?1<>0)";
+					case YN_BOOLEAN:
+						return "(?1<>'N')";
+					case TF_BOOLEAN:
+						return "(?1<>'F')";
+				}
+		}
```

Comment:
The only reservation I have here is that I guess there are some databases which will already natively do the right thing in some of these casts without the need for verbose `case`/`when` constructs.

Reply:
But it's _very_ nice that we can now cast between all the various different ways to represent booleans. That's actually rather useful.

Reply:
I thought about that as well, but if this turns out to be an issue, it could be optimized later. I wanted to keep it as simple as possible :)

Reply:
Well, actually we can only cast _from_ then to regular `Boolean`, because there's actually no way to specify things like `YesNoBooleanType` in the `cast()` function. So this is actually a bit less useful than it looks, unless we add a syntax for doing that.

So perhaps we should let you write something like `cast(foo as Boolean(T,F))` or something like that.